File: TestSample/csharp-akka/src/Actors/PersonActor.cs
using Akka.Actor;
using HelloAkka.Messages;

namespace HelloAkka.Actors;

/// <summary>
/// 모든 Speaker 액터의 추상 베이스. <see cref="Greeting"/>만 자식이 구현하면
/// 메시지 핸들링과 응답 패턴은 공통으로 처리된다.
/// </summary>
public abstract class PersonActor : ReceiveActor
{
    protected string Name { get; }
    protected string LanguageTag { get; }

    protected PersonActor(string name, string languageTag)
    {
        Name = name;
        LanguageTag = languageTag;

        Receive<SayHello>(_ =>
        {
            Sender.Tell(new HelloResponse(LanguageTag, Name, Greeting()));
        });
    }

    protected abstract string Greeting();
}

File: TestSample/java-akka/src/main/java/helloakka/actors/PersonActor.java
package helloakka.actors;

import akka.actor.AbstractActor;
import helloakka.messages.HelloResponse;
import helloakka.messages.SayHello;

/**
 * 모든 Speaker 액터의 추상 베이스. greeting() 만 자식이 구현하면
 * SayHello 메시지 처리/응답은 공통으로 처리된다.
 */
public abstract class PersonActor extends AbstractActor {
    protected final String name;
    protected final String languageTag;

    protected PersonActor(String name, String languageTag) {
        this.name = name;
        this.languageTag = languageTag;
    }

    protected abstract String greeting();

    @Override
    public Receive createReceive() {
        return receiveBuilder()
            .match(SayHello.class, msg ->
                getSender().tell(new HelloResponse(languageTag, name, greeting()), getSelf()))
            .build();
    }
}
